File: tasks/possibleHeights.py
def possibleHeights(parent):

    edges = [[] for i in range(len(parent))]
    height = [0 for i in range(len(parent))]
    isPossibleHeight = [False for i in range(len(parent))]

    def initGraph(parent):
        for i in range(1, len(parent)):
            edges[parent[i]].append(i)

    def calcHeight(v):
        for u in edges[v]:
            calcHeight(u)
            height[v] = max(height[v], height[u]+1)

        countHeights = [[] for i in range(len(edges))]
        for i in range(len(edges[v])):
            u = edges[v][i]
            countHeights[height[u]].append(u)
        edges[v] = []
        for i in range(len(edges) - 1, -1, -1):
            for j in range(len(countHeights[i])):
                edges[v].append(countHeights[i][j])

    def findNewHeights(v, tailHeight):
        isPossibleHeight[max(height[v], tailHeight)] = True
        firstMaxHeight = tailHeight + 1
        secondMaxHeight = tailHeight + 1
        if len(edges[v]) > 0:
            firstMaxHeight = max(firstMaxHeight, height[edges[v][0]] + 2)
        if len(edges[v]) > 1:
            secondMaxHeight = max(secondMaxHeight, height[edges[v][1]] + 2)
        if len(edges[v]) > 0:
            findNewHeights(edges[v][0], secondMaxHeight)
        for i in range(1, len(edges[v])):
            findNewHeights(edges[v][i], firstMaxHeight)

    initGraph(parent)
    calcHeight(0)
    findNewHeights(0, 0)

    heights = []
    for i in range(len(parent)):
        if isPossibleHeight[i]:
            heights.append(i)
    return heights

File: tasks/possibleHeights.js
function possibleHeights(parent) {

  var edges;
  var height;
  var isPossibleHeight;

  initGraph = function(parent) {
    edges = []
    height = []
    for (var i = 0; i < parent.length; i++) {
      edges.push([]);
      height.push(0);
    }
    for (var i = 1; i < parent.length; i++) {
      edges[parent[i]].push(i);
    }
    isPossibleHeight = []
  };

  var calcHeight = function(v) {
    for (var i = 0; i < edges[v].length; i++) {
      var u = edges[v][i];
      calcHeight(u);
      height[v] = Math.max(height[v], height[u] + 1);
    }

    countHeights = [];
    for (var i = 0; i < edges.length; i++) {
      countHeights[i] = [];
    }
    for (var i = 0; i < edges[v].length; i++) {
      var u = edges[v][i];
      countHeights[height[u]].push(u);
    }
    edges[v] = [];
    for (var i = edges.length - 1; i >= 0; i--) {
      for (var j = 0; j < countHeights[i].length; j++) {
        edges[v].push(countHeights[i][j]);
      }
    }
  };

  var findNewHeights = function(v, tailHeight) {
    isPossibleHeight[Math.max(height[v], tailHeight)] = true;
    var firstMaxHeight = tailHeight + 1;
    var secondMaxHeight = tailHeight + 1;
    if (edges[v].length > 0) {
      firstMaxHeight = Math.max(firstMaxHeight, height[edges[v][0]] + 2);
    }
    if (edges[v].length > 1) {
      secondMaxHeight = Math.max(secondMaxHeight, height[edges[v][1]] + 2);
    }
    if (edges[v].length > 0) {
      findNewHeights(edges[v][0], secondMaxHeight);
    }
    for (var i = 1; i < edges[v].length; i++) {
      findNewHeights(edges[v][i], firstMaxHeight);
    }
  }

  initGraph(parent);
  calcHeight(0);
  findNewHeights(0, 0);

  var result = [];
  for (var v in isPossibleHeight) {
    result.push(parseInt(v));
  }
  return result;
}
